File: lib/python/opentoken-cli/src/main/opentoken_cli/io/record_id_mapping_writer.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import csv
import logging
import os
from typing import Optional

logger = logging.getLogger(__name__)

MAPPING_FILE_SUFFIX = ".record-id-mapping.csv"
ORIGINAL_RECORD_ID = "original_record_id"
HASHED_RECORD_ID = "hashed_record_id"


class RecordIdMappingWriter:
    """
    Writes a record-ID mapping CSV file containing two columns:
    ``original_record_id`` and ``hashed_record_id``.

    Used when ``--hash-record-ids`` is specified on the ``tokenize`` or
    ``package`` subcommands, so that callers can reconcile hashed output back
    to their source records.
    """

    def __init__(self, file_path: str):
        """
        Open the mapping file for writing and emit the CSV header.

        Args:
            file_path: Path of the mapping CSV file to create.

        Raises:
            IOError: If the file cannot be opened for writing.
        """
        self.file_path = file_path
        os.makedirs(os.path.dirname(os.path.abspath(file_path)), exist_ok=True)
        self._file_handle = open(file_path, "w", newline="", encoding="utf-8")
        self._csv_writer = csv.writer(self._file_handle, lineterminator="\n")
        self._csv_writer.writerow([ORIGINAL_RECORD_ID, HASHED_RECORD_ID])

    def write_mapping(self, original_record_id: str, hashed_record_id: str) -> None:
        """
        Write a mapping row to the CSV file.

        Args:
            original_record_id: The original record ID.
            hashed_record_id: The SHA-256 hash of the record ID.
        """
        self._csv_writer.writerow([original_record_id, hashed_record_id])

    def close(self) -> None:
        """Close the underlying file handle."""
        if self._file_handle:
            self._file_handle.close()

    def __enter__(self) -> "RecordIdMappingWriter":
        return self

    def __exit__(self, exc_type, exc_val, exc_tb) -> None:
        self.close()

    @staticmethod
    def build_mapping_file_path(output_file_path: str) -> str:
        """
        Build the mapping file path from an output file path by stripping the
        file extension and appending ``MAPPING_FILE_SUFFIX``.

        Args:
            output_file_path: The token output file path.

        Returns:
            The corresponding mapping file path.
        """
        output_dir = os.path.dirname(output_file_path)
        base_name = os.path.splitext(os.path.basename(output_file_path))[0]
        return os.path.join(output_dir, base_name + MAPPING_FILE_SUFFIX)

File: lib/java/opentoken-cli/src/main/java/com/truveta/opentoken/cli/io/RecordIdMappingWriter.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.cli.io;

import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;

import org.apache.commons.csv.CSVFormat;
import org.apache.commons.csv.CSVPrinter;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Writes a record-ID mapping CSV file containing two columns:
 * {@code original_record_id} and {@code hashed_record_id}.
 *
 * <p>The file is used when {@code --hash-record-ids} is specified on the
 * {@code tokenize} or {@code package} subcommands, so that callers can
 * reconcile hashed output back to their source records.
 */
public class RecordIdMappingWriter implements AutoCloseable {

    public static final String ORIGINAL_RECORD_ID = "original_record_id";
    public static final String HASHED_RECORD_ID = "hashed_record_id";
    public static final String MAPPING_FILE_SUFFIX = ".record-id-mapping.csv";

    private static final Logger logger = LoggerFactory.getLogger(RecordIdMappingWriter.class);

    private final BufferedWriter fileWriter;
    private final CSVPrinter csvPrinter;
    private final String filePath;

    /**
     * Constructs a RecordIdMappingWriter for the given file path.
     *
     * @param filePath path of the mapping CSV file to create
     * @throws IOException if the file cannot be opened for writing
     */
    public RecordIdMappingWriter(String filePath) throws IOException {
        this.filePath = filePath;
        fileWriter = new BufferedWriter(new FileWriter(filePath));
        CSVFormat format = CSVFormat.Builder.create(CSVFormat.DEFAULT)
                .setRecordSeparator('\n')
                .setHeader(ORIGINAL_RECORD_ID, HASHED_RECORD_ID)
                .get();
        csvPrinter = new CSVPrinter(fileWriter, format);
    }

    /**
     * Writes a mapping row to the CSV file.
     *
     * @param originalRecordId the original record ID
     * @param hashedRecordId   the SHA-256 hash of the record ID
     * @throws IOException if an I/O error occurs
     */
    public void writeMapping(String originalRecordId, String hashedRecordId) throws IOException {
        csvPrinter.printRecord(originalRecordId, hashedRecordId);
    }

    /**
     * Returns the absolute path of the mapping file.
     *
     * @return file path
     */
    public String getFilePath() {
        return filePath;
    }

    /**
     * Builds the mapping file path from an output file path by stripping the
     * file extension and appending {@value #MAPPING_FILE_SUFFIX}.
     *
     * @param outputFilePath the token output file path
     * @return the corresponding mapping file path
     */
    public static String buildMappingFilePath(String outputFilePath) {
        int lastDotIndex = outputFilePath.lastIndexOf('.');
        String basePath = lastDotIndex > 0 ? outputFilePath.substring(0, lastDotIndex) : outputFilePath;
        return basePath + MAPPING_FILE_SUFFIX;
    }

    @Override
    public void close() throws IOException {
        try {
            csvPrinter.close();
        } catch (IOException e) {
            logger.warn("Error closing CSV printer for mapping file", e);
        }
        fileWriter.close();
    }
}
